pytorch_lightning/callbacks/base.py

```diff
@@ -34,20 +34,30 @@ class Callback(abc.ABC):
     """
 
     @property
-    def state_id(self) -> str:
+    def state_key(self) -> str:
         """
         Identifier for the state of the callback. Used to store and retrieve a callback's state from the
-        checkpoint dictionary by ``checkpoint["callbacks"][state_id]``. Implementations of a callback need to
-        provide a unique state id if 1) the callback has state and 2) it is desired to maintain the state of
+        checkpoint dictionary by ``checkpoint["callbacks"][state_key]``. Implementations of a callback need to
+        provide a unique state key if 1) the callback has state and 2) it is desired to maintain the state of
         multiple instances of that callback.
         """
         return self.__class__.__qualname__
 
     @property
-    def _legacy_state_id(self) -> Type["Callback"]:
-        """State identifier for checkpoints saved prior to version 1.5.0."""
+    def _legacy_state_key(self) -> Type["Callback"]:
+        """State key for checkpoints saved prior to version 1.5.0."""
         return type(self)
 
+    def _generate_state_key(self, **kwargs: Any) -> str:
```

Comment:
should this be an ordered container?

Reply:
I believe it is. In Python 3.6 and higher the kwargs and dict is insertion ordered. Is there another concern?

Reply:
I see, that was my main concern for the key to be consistent

Reply:
> In Python 3.6 and higher the kwargs and dict is insertion ordered. Is there another concern?

Actually, it's 3.7 and higher so this could be a problem for 3.6

https://mail.python.org/pipermail/python-dev/2017-December/151283.html

Reply:
so we are forced to use OrderedDict? I don't understand Gido's email.

stackoverflow answer says 3.6 has insertion order maintained: https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6

we should be fine relying on insertion order, right?

Reply:
No, it says that 3.6 has insertion order maintained due to undefined behavior.

The behaviour becomes defined (officially supported) in 3.7.

So we are technically forced to use `OrderedDict`

Reply:
nvm, PEP468 made it to 3.6 so `**kwargs` ARE ordered

https://docs.python.org/3/whatsnew/3.6.html#pep-468-preserving-keyword-argument-order